File: interviews/leetcode/leetcode-75/two-pointers/11. Container With Most Water/solution.go
package main

func maxArea(height []int) int {
	// initialize two pointers at the start and end of the array
	left := 0
	right := len(height) - 1
	maxArea := 0

	for left < right {
		// Calculate the current area
		width := right - left
		minHeight := height[left]
		if height[right] < height[left] {
			minHeight = height[right]
		}
		currentArea := minHeight * width

		// Update the maximum area
		if currentArea > maxArea {
			maxArea = currentArea
		}

		// Move the pointer for the shorter line
		if height[left] < height[right] {
			left++
		} else {
			right--
		}
	}

	return maxArea
}

/*
func main() {
    height := []int{1, 8, 6, 2, 5, 4, 8, 3, 7}
    fmt.Println(maxArea(height))  // Output: 49
}
*/

File: interviews/leetcode/leetcode-75/two-pointers/11. Container With Most Water/solution.py
class Solution:
    def maxArea(self, height: list[int]) -> int:
        left = 0  # start this pointer from the left side
        right = len(height) - 1 # start this pointer from the right side
        max_area = 0

        while left < right:
            # Calculate the current area
            current_area = min(height[left], height[right]) * (right - left)
            # Update the maximum area
            max_area = max(max_area, current_area)

            # Move the pointer for the shorter line
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        return max_area
